File: test_code/weather_forecast.py
import time
import signal
import requests
from datetime import datetime
from threading import Event

# Weather code mapping with icons
WEATHER_MAPPING = {
    0: ("Clear sky", "☀️"),
    1: ("Mainly clear", "🌤️"),
    2: ("Partly cloudy", "⛅"),
    3: ("Overcast", "☁️"),
    45: ("Fog", "🌫️"),
    48: ("Depositing rime fog", "🌫️"),
    51: ("Drizzle: Light", "🌦️"),
    53: ("Drizzle: Moderate", "🌦️"),
    55: ("Drizzle: Dense", "🌧️"),
    56: ("Freezing Drizzle: Light", "🌧️"),
    57: ("Freezing Drizzle: Dense", "🌧️"),
    61: ("Rain: Slight", "🌧️"),
    63: ("Rain: Moderate", "🌧️"),
    65: ("Rain: Heavy", "🌧️"),
    66: ("Freezing Rain: Light", "🌧️"),
    67: ("Freezing Rain: Heavy", "🌧️"),
    71: ("Snowfall: Slight", "❄️"),
    73: ("Snowfall: Moderate", "❄️"),
    75: ("Snowfall: Heavy", "❄️"),
    77: ("Snow grains", "❄️"),
    80: ("Rain showers: Slight", "🌦️"),
    81: ("Rain showers: Moderate", "🌧️"),
    82: ("Rain showers: Violent", "🌧️"),
    85: ("Snow showers: Slight", "❄️"),
    86: ("Snow showers: Heavy", "❄️"),
    95: ("Thunderstorm: Slight", "⛈️"),
    96: ("Thunderstorm with slight hail", "⛈️"),
    99: ("Thunderstorm with heavy hail", "⛈️")
}

def get_weather():
    LAT = 25.0585178  # Latitude from Google Maps
    LON = 121.6532539  # Longitude from Google Maps
    URL = (f"https://api.open-meteo.com/v1/forecast?latitude={LAT}&longitude={LON}" 
           f"&hourly=temperature_2m,relative_humidity_2m,precipitation_probability,weathercode"
           f"&timezone=Asia/Taipei")
    
    try:
        response = requests.get(URL)
        data = response.json()
        
        if "hourly" not in data:
            print("[ERROR] Unable to fetch weather data:", data)
            return None, None
        
        now = datetime.now().strftime("%Y-%m-%dT%H:00")
        if now in data["hourly"]["time"]:
            start_index = data["hourly"]["time"].index(now)
        else:
            print("[WARNING] Current time not found in weather data, using first available hour.")
            start_index = 0
        
        current_weather = {
            "temperature": data["hourly"]["temperature_2m"][start_index],
            "humidity": data["hourly"]["relative_humidity_2m"][start_index],
            "precipitation": data["hourly"]["precipitation_probability"][start_index],
            "weather_code": data["hourly"]["weathercode"][start_index],
            "time": data["hourly"]["time"][start_index]
        }
        
        forecast = []
        for i in range(start_index, start_index + 12):
            if i < len(data["hourly"]["time"]):
                forecast.append({
                    "time": data["hourly"]["time"][i],
                    "temperature": data["hourly"]["temperature_2m"][i],
                    "humidity": data["hourly"]["relative_humidity_2m"][i],
                    "precipitation": data["hourly"]["precipitation_probability"][i],
                    "weather_code": data["hourly"]["weathercode"][i]
                })
        
        return current_weather, forecast
    except Exception as e:
        print(f"[ERROR] Failed to retrieve weather data: {e}")
        return None, None

def signal_handler(signum, frame):
    print("\n[INFO] Received termination signal, exiting...")
    exit_event.set()

exit_event = Event()
signal.signal(signal.SIGINT, signal_handler)
signal.signal(signal.SIGTERM, signal_handler)

current_weather, forecast = get_weather()
next_weather_update = time.time() - 1

while not exit_event.is_set():
    now = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    print(f"\r[Current Time] {now}", end="", flush=True)
    
    if time.time() >= next_weather_update:
        print("\n[INFO] Updating weather data...")
        current_weather, forecast = get_weather()
        next_weather_update = time.time() + 3600
        
        if current_weather:
            description, icon = WEATHER_MAPPING.get(current_weather['weather_code'], ("Unknown", "❓"))
            print(f"[Current Weather] {current_weather['time']}, {current_weather['temperature']}°C, "
                  f"Humidity: {current_weather['humidity']}%, Precipitation: {current_weather['precipitation']}%, "
                  f"Weather: {description} {icon}")
        if forecast:
            print("[Next 12 Hours Forecast]:")
            for entry in forecast:
                description, icon = WEATHER_MAPPING.get(entry['weather_code'], ("Unknown", "❓"))
                print(f"  {entry['time']}: {entry['temperature']}°C, Humidity: {entry['humidity']}%, "
                      f"Precipitation: {entry['precipitation']}%, Weather: {description} {icon}")
    
    time.sleep(1)

File: test_code/weather_forecast.cpp
#include <iostream>
#include <iomanip>
#include <ctime>
#include <csignal>
#include <thread>
#include <chrono>
#include <curl/curl.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;
using namespace std;

volatile sig_atomic_t exit_flag = 0;

void signal_handler(int signum) {
    cout << "\n[INFO] Received signal: " << signum << " (";
    switch (signum) {
        case SIGINT: cout << "SIGINT"; break;
        case SIGTERM: cout << "SIGTERM"; break;
        default: cout << "Unknown"; break;
    }
    cout << ")\n";
    exit_flag = 1;
}

size_t write_callback(void* contents, size_t size, size_t nmemb, string* output) {
    size_t total_size = size * nmemb;
    output->append((char*)contents, total_size);
    return total_size;
}

json get_weather() {
    const string url = "https://api.open-meteo.com/v1/forecast?latitude=25.0585178&longitude=121.6532539" 
                       "&hourly=temperature_2m,relative_humidity_2m,precipitation_probability,weathercode"
                       "&timezone=Asia/Taipei";
    
    CURL* curl = curl_easy_init();
    if (!curl) {
        cerr << "[ERROR] Failed to initialize CURL" << endl;
        return json();
    }
    
    string response;
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
    
    CURLcode res = curl_easy_perform(curl);
    curl_easy_cleanup(curl);
    
    if (res != CURLE_OK) {
        cerr << "[ERROR] Failed to fetch weather data: " << curl_easy_strerror(res) << endl;
        return json();
    }
    
    return json::parse(response, nullptr, false);
}

string format_time(time_t raw_time) {
    struct tm* time_info = localtime(&raw_time);
    stringstream ss;
    ss << put_time(time_info, "%Y/%b/%d %H:%M:%S");
    return ss.str();
}

string get_weather_description(int code) {
    switch (code) {
        case 0: return "Clear sky";
        case 1: case 2: case 3: return "Partly cloudy";
        case 45: case 48: return "Fog";
        case 51: case 53: case 55: return "Drizzle";
        case 61: case 63: case 65: return "Rain";
        case 71: case 73: case 75: return "Snow";
        case 80: case 81: case 82: return "Rain showers";
        case 95: case 96: case 99: return "Thunderstorm";
        default: return "Unknown";
    }
}

int main() {
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    
    time_t next_weather_update = time(nullptr) - 1;
    json weather_data;
    
    while (!exit_flag) {
        time_t now = time(nullptr);
        cout << "\r[Current Time] " << format_time(now) << flush;
        
        if (now >= next_weather_update) {
            cout << "\n[INFO] Updating weather data..." << endl;
            weather_data = get_weather();
            next_weather_update = now + 3600;
            
            if (!weather_data.empty() && weather_data.contains("hourly")) {
                auto& hourly = weather_data["hourly"];
                vector<string> times = hourly["time"].get<vector<string>>();
                size_t index = 0;
                while (index < times.size()) {
                    struct tm tm = {};
                    strptime(times[index].c_str(), "%Y-%m-%dT%H:%M", &tm);
                    time_t forecast_time = mktime(&tm);
                    if (forecast_time >= now) break;
                    ++index;
                }
                
                cout << "[Next 12 Hours Forecast]:" << endl;
                for (size_t i = 0; i < 12 && (index + i) < times.size(); ++i) {
                    size_t idx = index + i;
                    int weather_code = hourly["weathercode"][idx];
                    cout << "  " << times[idx] << ": "
                         << hourly["temperature_2m"][idx] << "°C, "
                         << "Humidity: " << hourly["relative_humidity_2m"][idx] << "%, "
                         << "Precipitation: " << hourly["precipitation_probability"][idx] << "%, "
                         << "Condition: " << get_weather_description(weather_code) << endl;
                }
            }
        }
        
        this_thread::sleep_for(chrono::seconds(1));
    }
    
    cout << "\n[INFO] Exiting..." << endl;
    return 0;
}

